R/contribution_matrix.R

```diff
@@ -0,0 +1,588 @@
+      #Populate the middle second column of X_star
+      X_star[(nrow(X) + 1):(2 * nrow(X)), (nrow(X) + 1):(nrow(X) + ncol(X_d))] <- -X_d
+      #Populate the middle third column of X_star
+      X_star[(nrow(X) + 1):(2 * nrow(X)), (nrow(X) + ncol(X_d) + 1):(nrow(X) + 2 * ncol(X_d))] <- -X_beta
+      #Populate the bottom third column of X_star
+      X_star[(2 * nrow(X) + 1):nrow(X_star), (nrow(X) + ncol(X_d) + 1):(nrow(X) + 2 * ncol(X_d))] <- diag(1, nrow = ncol(X_d))
+      #Populate the bottom right column of X_star
+      X_star[(2 * nrow(X) + 1):nrow(X_star), (nrow(X) + 2 * ncol(X_d) + 1):ncol(X_star)] <- -matrix(1, nrow = ncol(X_d), ncol = 1)
+      
+      A <- solve(t(X_star) %*% solve(V_star) %*% X_star) %*% t(X_star) %*% solve(V_star)
+      A_row2_left <- A[(ncol(X) + 1):(ncol(X) + ncol(X_d)), 1:nrow(X)]
+      A_row3_left <- A[(ncol(X) + ncol(X_d) + 1):(ncol(X) + 2 * ncol(X_d)), 1:nrow(X)]
+      
+      if(all_parameters){
+        contribution <- Z %*% rbind(A_row2_left, A_row3_left)
+      } else if(!all_parameters){
+        contribution <- rbind(A_row2_left, A_row3_left)
+      }
+    }  
+    
+  } else{
+    stop("effects_type must be 'Fixed' or 'Random'")
+  }
+  
+  contribution_abs <- abs(contribution)
+  contribution_row_sums <- rowSums(contribution_abs)
+  contribution_percent <- 100 * contribution_abs / contribution_row_sums
+  
+  #Select percentage contributions or absolute contributions
+  if(percentages){
+    contribution_output <- t(contribution_percent)
+  } else if(!percentages){
+    contribution_output <- t(contribution_abs)
+  }
+  
+  #Create a study level contribution matrix if required
+  if(study_level){
+    #Number of studies
+    n_studies <- length(studies)
+    #Number of arms per study
+    n_arms <- table(data$Study)
+    
+    #Local function to extract the studies from a vector of strings that are in the format "(Study)Control_treatment:Treatment2"
+    GetStudies <- function(x){
+      substr(x, start = 2, stop = unlist(gregexpr(")", x)) - 1)
+    }
+    
+    #Rename rows by study only
+    rownames(contribution_output) <- GetStudies(rownames(contribution_output))
+    #Create the contribution matrix for studies only with the correct dimensions
+    contribution_study <- matrix(nrow = n_studies, ncol = ncol(contribution_output))
+    rownames(contribution_study) <- studies
+    colnames(contribution_study) <- colnames(contribution_output)
+    for(study in studies){
+      #If there is only one row in the contribution matrix for this study, copy it
+      if(n_arms[study] == 2){
+        contribution_study[which(rownames(contribution_study) == study), ] <- contribution_output[which(rownames(contribution_output) == study), ]
+      #If there is more than one row for this study, add up the rows
+      } else if(n_arms[study] > 2){
+        contribution_study[which(rownames(contribution_study) == study), ] <- colSums(contribution_output[which(rownames(contribution_output) == study), ])
```

Comment:
I don't know what you mean. Can you explain further?

Reply:
Say somebody wanted a contribution matrix at the study level made up of percentages where there are some studies with more than 2 arms.
From what I can see, first the percentages are calculated across each comparison. Then, if the user has selected 'study level', for studies that had multiple arms, the percentages are summed (column wise). Naturally this is fine for absolute values, but for percentages that would mean that they would add up to more than 100% (i.e. it would be 200% for 3-arm studies, 300% for 4-arm studies...)?
I think if it did the study level thing first (i.e. summed via columns for studies with multiple arms) and then did the percentage thing it would be fine. But with the way round it currently is, I think you'd need to divide the percentages by (number of arms - 1).

Reply:
I think the confusion here is because the matrix is transposed (somewhat arbitrarily as per a previous comment). The percentages are by row (meaning each row sums to 100) before transposing. After transposing the percentages are by column. So the crucial thing is whether the study level matrix is created before or after transposing. Line 680 is where the percentage matrix is transposed, before the study level matrix is created. Do we agree now?

Reply:
Thanks for the clarification. I turns out that I got confused with essentially what the column headers were for the contribution matrix (before transposition) and though they were already at the study level, and so thought we were adding stuff together in a different dimension.

Anyway - I'm happy now - thanks :)